### PR TITLE
#697: Add support for HTTP GET calls for GetCapabilities & DescribeProcess.

### PR DESCRIPTION
**Pull request by ccmoloney**

### CODE REVIEW

File: wps-common/src/main/resources/xml/GetCapabilities/GetCapabilities.ftl [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+                <ows:HTTP>
+                    <ows:Get xlink:href="${geoserverWPSEndpointURL}"/>
+                    <ows:Post xlink:href="${geoserverWPSEndpointURL}"/>
+                </ows:HTTP>
+            </ows:DCP>
+        </ows:Operation>
+        <ows:Operation name="Execute">
+            <ows:DCP>
+                <ows:HTTP>
+                    <ows:Get xlink:href="${geoserverWPSEndpointURL}"/>
```
Comment (jonescc) [inline wps-common/src/main/resources/xml/GetCapabilities/GetCapabilities.ftl]: Not supported?  Remove?

File: wps-common/src/main/resources/xml/GetCapabilities/GetCapabilities.ftl [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+                    <ows:Post xlink:href="${geoserverWPSEndpointURL}"/>
+                </ows:HTTP>
+            </ows:DCP>
+        </ows:Operation>
+    </ows:OperationsMetadata>
+    <wps:ProcessOfferings>
+        <wps:Process wps:processVersion="1.0.0">
```
Comment (jonescc) [inline wps-common/src/main/resources/xml/GetCapabilities/GetCapabilities.ftl]: Should remove everything here except for gogoduck and possibly netcdfoutput for the moment.   The others won't be supported - they were geoserver processes.

File: request-handler/src/main/java/au/org/aodn/aws/wps/QueryStringParameterParser.java [hunk 3]
```diff
@@ -1,18 +1,100 @@
...
+            Operation operation = OperationFactory.getInstance(request);
+
+            LOGGER.info("Operation returned: " + operation.getClass().getName());
+
+            if (operation instanceof DescribeProcessOperation)
+            {
+                //  Set relevant parameters
+                DescribeProcessOperation describeOperation = (DescribeProcessOperation) operation;
+
+                DescribeProcess describeRequest = describeOperation.getRequest();
```
Comment (jonescc) [inline request-handler/src/main/java/au/org/aodn/aws/wps/QueryStringParameterParser.java]: Looks to me like you actually wanted to the OGC request above and then the operation (using the existing getOperation in OperationFactory)

File: request-handler/src/main/java/au/org/aodn/aws/wps/QueryStringParameterParser.java [hunk 4]
```diff
@@ -1,18 +1,100 @@
...
     public Operation getOperation() {
-        throw new UnsupportedOperationException("Not yet implemented");
+        if(queryParameters.get(REQUEST_NAME_PARAMETER_NAME) != null)
+        {
+
+            String request = queryParameters.get(REQUEST_NAME_PARAMETER_NAME);
+
+            LOGGER.info("Request name: " + request);
+
+            Operation operation = OperationFactory.getInstance(request);
```
Comment (jonescc) [inline request-handler/src/main/java/au/org/aodn/aws/wps/QueryStringParameterParser.java]: Perhaps use a HTTP Get RequestFactory/Parser to get the OGC request and then use the existing OperationFactory getOperation method to get the operation - would read better to me and wouldn't require the getRequest method on operation.

File: request-handler/src/main/java/au/org/aodn/aws/wps/operation/DescribeProcessOperation.java [hunk 5]
```diff
@@ -34,52 +36,86 @@ public DescribeProcessOperation(DescribeProcess request) {
...
+        return this.request;
+    }
+
     @Override
     public String execute()
     {
-        String processDescriptionsS3Bucket = WpsConfig.getConfig(DESCRIBE_PROCESS_S3_BUCKET_CONFIG_KEY);
-        String S3KeyPrefix = WpsConfig.getConfig(DESCRIBE_PROCESS_S3_KEY_PREFIX_CONFIG_KEY);
-        String s3RegionName = WpsConfig.getConfig(AWS_REGION_CONFIG_KEY);
+        String processDescriptionsS3Bucket = WpsConfig.getConfig(WpsConfig.DESCRIBE_PROCESS_S3_BUCKET_CONFIG_KEY);
```
Comment (jonescc) [inline request-handler/src/main/java/au/org/aodn/aws/wps/operation/DescribeProcessOperation.java]: Would a static import make these constant references cleaner/easier to read?

File: wps-common/src/main/resources/xml/DescribeProcess/GoGoDuck.xml [hunk 6]
```diff
@@ -0,0 +1,69 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<wps:ProcessDescriptions xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:wps="http://www.opengis.net/wps/1.0.0"
```
Comment (jonescc) [inline wps-common/src/main/resources/xml/DescribeProcess/GoGoDuck.xml]: Can we just include the ProcessDescription element?  The ProcessDescriptions elelemtn is superfluous.

File: wps-common/src/main/resources/xml/DescribeProcess/NetcdfOutput.xml [hunk 7]
```diff
@@ -0,0 +1,64 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<wps:ProcessDescriptions xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:wps="http://www.opengis.net/wps/1.0.0"
```
Comment (jonescc) [inline wps-common/src/main/resources/xml/DescribeProcess/NetcdfOutput.xml]: And here.

File: wps-common/src/main/resources/xml/GetCapabilities/GetCapabilities.ftl (hunk 1 above)
Comment (ccmoloney) [reply wps-common/src/main/resources/xml/GetCapabilities/GetCapabilities.ftl]: Will remove.

File: wps-common/src/main/resources/xml/GetCapabilities/GetCapabilities.ftl (hunk 2 above)
Comment (ccmoloney) [reply wps-common/src/main/resources/xml/GetCapabilities/GetCapabilities.ftl]: Will remove.

File: wps-common/src/main/resources/xml/DescribeProcess/GoGoDuck.xml (hunk 6 above)
Comment (ccmoloney) [reply wps-common/src/main/resources/xml/DescribeProcess/GoGoDuck.xml]: I tried that exact method - but got XML unmarshalling issues - so went down this path.

File: wps-common/src/main/resources/xml/GetCapabilities/GetCapabilities.ftl [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+        </wps:Process>
+        <wps:Process wps:processVersion="1.0.0">
+            <ows:Identifier>gs:StoreCoverage</ows:Identifier>
+            <ows:Title>Store Coverage</ows:Title>
+            <ows:Abstract>Stores a raster on the server.</ows:Abstract>
+        </wps:Process>
+    </wps:ProcessOfferings>
+    <wps:Languages>
+        <wps:Default>
+            <ows:Language>en-US</ows:Language>
```
[the remark this answers is not in the document]
Comment (ccmoloney) [reply wps-common/src/main/resources/xml/GetCapabilities/GetCapabilities.ftl]: Will amend code to reflect the language specified here.

File: request-handler/src/main/java/au/org/aodn/aws/wps/QueryStringParameterParser.java [hunk 9]
```diff
@@ -1,18 +1,100 @@
...
         this.queryParameters = request.getQueryStringParameters();
     }
 
     @Override
     public Operation getOperation() {
-        throw new UnsupportedOperationException("Not yet implemented");
+        if(queryParameters.get(REQUEST_NAME_PARAMETER_NAME) != null)
+        {
+
+            String request = queryParameters.get(REQUEST_NAME_PARAMETER_NAME);
```
[the remark this answers is not in the document]
Comment (ccmoloney) [reply request-handler/src/main/java/au/org/aodn/aws/wps/QueryStringParameterParser.java]: Will cater for upper or lowercase parameters.

File: request-handler/src/main/java/au/org/aodn/aws/wps/QueryStringParameterParser.java (hunk 4 above)
Comment (ccmoloney) [reply request-handler/src/main/java/au/org/aodn/aws/wps/QueryStringParameterParser.java]: Will do in a subsequent pull request.

File: request-handler/src/main/java/au/org/aodn/aws/wps/QueryStringParameterParser.java [hunk 10]
```diff
@@ -97,4 +99,19 @@ else if (operation instanceof GetCapabilitiesOperation)
         //  TODO: exception?
         return null;
     }
+
+
+    private String getMapValueIgnoreCase(String key, Map<String, String> map)
+    {
+        if(key != null && map != null) {
+            if (map.get(key.toLowerCase()) != null) {
```
Comment (jonescc) [inline request-handler/src/main/java/au/org/aodn/aws/wps/QueryStringParameterParser.java]: Mixed case is also OK e.g. 'Request' - perhaps just loop through the keys and find one which matches when converted to lower case

File: request-handler/src/main/java/au/org/aodn/aws/wps/QueryStringParameterParser.java (hunk 3 above)
Comment (ccmoloney) [reply request-handler/src/main/java/au/org/aodn/aws/wps/QueryStringParameterParser.java]: Will revisit.  This is working currently + will do for this iteration.

File: request-handler/src/main/java/au/org/aodn/aws/wps/operation/DescribeProcessOperation.java (hunk 5 above)
Comment (ccmoloney) [reply request-handler/src/main/java/au/org/aodn/aws/wps/operation/DescribeProcessOperation.java]: Potato/Potato  :)

File: request-handler/src/main/java/au/org/aodn/aws/wps/QueryStringParameterParser.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (ccmoloney) [reply request-handler/src/main/java/au/org/aodn/aws/wps/QueryStringParameterParser.java]: Fixed.  Upper, lower and mixed case recognised.

**Pull request merged**